### ISSUE TITLE
data trace: type error should not show trace

### ISSUE THREAD

**Issue by sanderr** [inmanta/inmanta#2013]
Compiling the model
```
entity A:
    number n
end

implement A using std::none

x = A(n = "test")
```

results in the following output:

```
Could not set attribute `n` on instance `__config__::A (instantiated at ./main.cf:7)`
data trace:
attribute n on __config__::A instance
SUBTREE for __config__::A instance:
    CONSTRUCTED BY `A(n='test')`
    AT ./main.cf:7
└── 'test'
    SET BY `A(n='test')`
    AT ./main.cf:7
 (reported in Construct(A) (./main.cf:7))
caused by:
  Invalid value 'test', expected Number (reported in Construct(A) (./main.cf:7))
```

The trace is unhelpful in this case since it's a type error, not a double assignment error.

**Comment by sanderr** [comment]
Actually I believe the data trace might be helpful in this case as well. Take for example a slight deviation from the model above:
```
entity A:
    number n
end

implement A using std::none

x = A(n = m)

n = m
m = l
l = k
k = "test"
```
Which produces
```
Could not set attribute `n` on instance `__config__::A (instantiated at ./test.cf:7)`
data trace:
attribute n on __config__::A instance
SUBTREE for __config__::A instance:
    CONSTRUCTED BY `A(n=m)`
    AT ./test.cf:7
└── m
    SET BY `A(n=m)`
    AT ./test.cf:7
    └── l
        SET BY `m = l`
        AT ./test.cf:10
        └── k
            SET BY `l = k`
            AT ./test.cf:11
            └── 'test'
                SET BY `k = 'test'`
                AT ./test.cf:12
 (reported in Construct(A) (./test.cf:7))
caused by:
  Invalid value 'test', expected Number (reported in Construct(A) (./test.cf:7))
```
The purpose of the data trace is in all cases to give more insight in the flow of data. It does that for this case. Admittedly, cases such as the first example model might be the norm, but overall I'm not sure the data trace should be disabled for type errors.
@wouterdb @bartv what do you think?

**Comment by bartv** [comment]
true, but maybe the issue is that you first get the error, then the trace and then the caused by.

In this case the caused by information is really important to understand the datatrace

**Comment by sanderr** [comment]
That's a valid point. I'm not sure yet how that could best be resolved, but I'll look into it.

**Comment by sanderr** [comment]
I've had another look at it and I'd say #1960 is a prerequisite for solving this cleanly. The current approach is to look for the deepest (most specific error cause) for which a trace can be generated and attach it to that exception. In this case, the type exception itself does not warrant a trace, but the attribute exception wrapping it does. So the trace is added to the end of the attribute exception's message. More advanced exception extensions are difficult without the added abstractions that should be provided by #1960.
I propose to set #1960 as a prerequisite for this and to either move this ticket to backlog, or move #1960 to next.
